syntax = "proto3";
option optimize_for = SPEED;
package XStore_Proto;

// ======== COMMAND directives section ========
enum requestType {
  INFO = 0;
  CREATEDB = 1;
  REMOVEDB = 2;
  REMOVEDBFILES = 3;
  UNARYINSERT = 4;
  UNARYQUERY = 5;
  UNARYDELETE = 6;
  BATCHINSERT = 7;
  BATCHQUERY = 8;
  BATCHDELETE = 9;
  RANGEINSERT = 10;
  RANGEQUERY = 11;
  RANGEDELETE = 12;
  MIN = 13;
  MAX = 14;
  SUM = 15;
  AVG = 16;
}

enum timeUnit {
  SECOND = 0;
  MILLISECOND = 1;
  MICROSECOND = 2;
  NANOSECOND = 3;
}

message configStruct {
  string dbName = 1;
  timeUnit timeResolution = 2;
  sfixed32 rowLength = 3;
}

message fData {
  oneof oneOfData {
    double doubleData = 1;
    float floatData = 2;
    int32 int32Data = 3;
    int64 int64Data = 4;
    uint32 uint32Data = 5;
    uint64 uint64Data = 6;
    sint32 sint32Data = 7;
    sint64 sint64Data = 8;
    fixed32 fixed32Data = 9;
    fixed64 fixed64Data = 10;
    sfixed32 sfixed32Data = 11;
    sfixed64 sfixed64Data = 12;
    bool boolData = 13;
    string stringData = 14;
    bytes bytesData = 15;
  }
}

message rowData {
  repeated fData fieldData = 1;
}

message multipleRowData {
  repeated rowData rData = 1;
}

message reqCMD {
  requestType rType = 1;
}

// ======== REQUESTS section ========
// --- CREATE DB request
message createDB_REQ {
  requestType rType = 1;
  string dbName = 2;
  timeUnit timeResolution = 3;
  uint32 rowLength = 4;
}

// --- REMOVE DB request
message removeDB_REQ {
  requestType rType = 1;
  string dbName = 2;
}

// --- REMOVE DB FILES request
message removeDBFiles_REQ {
  requestType rType = 1;
  string dbName = 2;
  repeated string fileName = 3;
}

// --- UNARY INSERT
message unaryInsert_REQ {
  requestType rType = 1;
  uint64 timestamp = 2;
  rowData data = 3;
  string dbName = 4;
}

// --- UNARY QUERY
message unaryQuery_REQ {
  requestType rType = 1;
  uint64 timestamp = 2;
  string dbName = 3;
}

// --- UNARY DELETE
message unaryDelete_REQ {
  requestType rType = 1;
  uint64 timestamp = 2;
  string dbName = 3;
}

// --- BATCH INSERT (DATA)
message batchInsert_DATA {
  uint64 timestamp = 1;
  rowData data = 2;
}

// --- BATCH INSERT
message batchInsert_REQ {
  requestType rType = 1;
  string dbName = 2;
  repeated batchInsert_DATA batchInsert_DATA = 3;
}

// --- BATCH QUERY
message batchQuery_REQ {
  requestType rType = 1;
  repeated uint64 timestamp = 2;
  string dbName = 3;
}

// --- BATCH DELETE
message batchDelete_REQ {
  requestType rType = 1;
  repeated uint64 timestamp = 2;
  string dbName = 3;
}

// --- RANGE INSERT (DATA)
message rangeInsert_DATA {
  uint64 timestamp = 1;
  rowData data = 2;
}

// --- RANGE INSERT
message rangeInsert_REQ {
  requestType rType = 1;
  string dbName = 2;
  repeated rangeInsert_DATA rangeInsert_DATA = 3;
}

// --- RANGE QUERY
message rangeQuery_REQ {
  requestType rType = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
  string dbName = 4;
}

// --- RANGE DELETE
message rangeDelete_REQ {
  requestType rType = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
  string dbName = 4;
}

// --- MIN
message min_REQ {
  requestType rType = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
  string dbName = 4;
  sint32 columnIdx = 5;
}

// --- MAX
message max_REQ {
  requestType rType = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
  string dbName = 4;
  sint32 columnIdx = 5;
}

// --- SUM
message sum_REQ {
  requestType rType = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
  string dbName = 4;
  sint32 columnIdx = 5;
}

// --- AVG
message avg_REQ {
  requestType rType = 1;
  uint64 startTime = 2;
  uint64 endTime = 3;
  string dbName = 4;
  sint32 columnIdx = 5;
}

// ======== RESPONSE section ========
// --- DB Stats & Info REP
message dbStats {
  string dbName = 1;
  timeUnit dbResolution = 2;
  uint64 totalSizeInByte = 3;
  repeated string fileName = 4;
}

// --- SYS INFO
message sysInfo_REP {
  string serverAddr = 1;
  int32 threadCount = 2;
  repeated dbStats dbInfo = 3;
}

// --- CREATE DB response
message createDB_REP {
  string createDBStatus = 1;
}

// --- REMOVE DB
message removeDB_REP {
  string removeStatus = 1;
}

// --- REMOVE DB FILES
message removeDBFiles_REP {
  repeated string removeStatus = 1;
}

// --- UNARY INSERT
message unaryInsert_REP {
  string insertStatus = 1;
}

// --- UNARY QUERY
message unaryQuery_REP {
  uint64 timestamp = 1;
  rowData data = 2;
  optional string queryStatus = 3;
}

// --- UNARY DELETE
message unaryDelete_REP {
  string deleteStatus = 1;
}

// --- BATCH INSERT
message batchInsert_REP {
  string insertStatus = 1;
}

// --- BATCH QUERY
message batchQuery_REP {
  repeated unaryQuery_REP queryResponse = 1;
  optional string queryStatus = 3;
}

// --- BATCH DELETE
message batchDelete_REP {
  string deleteStatus = 1;
}

// --- RANGE INSERT
message rangeInsert_REP {
  string insertStatus = 1;
}

// --- RANGE QUERY
message rangeQuery_REP {
  repeated unaryQuery_REP queryResponse = 1;
  optional string queryStatus = 3;
}

// --- RANGE DELETE
message rangeDelete_REP {
  string deleteStatus = 1;
}

// --- MIN
message min_REP {
  fData opResult = 1;
  string opStatus = 2;
}

// --- MAX
message max_REP {
  fData opResult = 1;
  string opStatus = 2;
}

// --- SUM
message sum_REP {
  double opResult = 1;
  string opStatus = 2;
}

// --- AVG
message avg_REP {
  double opResult = 1;
  string opStatus = 2;
}